storage/innobase/row/row0ins.cc

```diff
@@ -2831,6 +2831,7 @@ row_ins_clust_index_entry_low(
 		/* If foreign key exist and foreign key is enabled
 		then avoid using bulk insert for copy algorithm */
 		if (innodb_alter_copy_bulk
+		    && !index->table->unique_blob
 		    && !index->table->is_temporary()
 		    && !index->table->versioned()
 		    && !index->table->has_spatial_index()
```

Comment:
Instead of relying on a new redundant field of `dict_table_t`, can we move all these checks to a new member function that would defined in this file:
```c++
inline bool dict_table_t::can_bulk_insert(const trx_t &trx) noexcept
{
  if (is_temporary() || versioned() || has_spatial_index())
    return false;
  if (n_v_cols /* && DB_xxx exists in v_col_names */)
    return false;
  return !trx.check_foreigns || (foreign_set.empty() && referenced_set.empty());
}
```
The condition in the caller would then just be the following:
```c++
		if (innodb_alter_copy_bulk
		    && index->table->can_bulk_insert(*trx)) {
```

Reply:
Before inserting into the engine, they do search the key on the index. If they find the matching value then they abort. But InnoDB bulk insert does insert at the end of all alter copy. So this has to be skipped when bulk insert is involved. 

Virtual index is being handled well by bulk operation. Because row passed to engine has all virtual column value as well.